File: C#/bob/Bob.cs
using System;
using System.Linq;

public static class Bob
{
    [Flags]
    public enum StatementAttributes
    {
        Empty =     1 << 0,   //  1 or 00000001     
        Question =  1 << 1,   //  2 or 00000010     
        AllCaps =   1 << 2,   //  4 or 00000100     
        Unused1 = 0,
        Unused2 = 0,
        Unused3 = 0,
        Unused4 = 0,
        AllOn =     7         //  7 or 00000111
    }

    public static string Response(string statement)
    {
        // create new field to hold attributes for a statement
        StatementAttributes attributes = StatementAttributes.Unused1;
        bool isEmpty = IsEmpty(statement);
        bool isQuestion = IsQuestion(statement);
        bool isAllCaps = !isEmpty && IsAllCaps(statement);

        // Check for attributes to turn on 
        if (isEmpty)
        {
            attributes |= StatementAttributes.Empty;        // Turn on the 'empty' field 
        }

        if (isQuestion)
        {
            attributes |= StatementAttributes.Question;     // Turn on the 'question' field 
        }

        if (isAllCaps)
        {
            attributes |= StatementAttributes.AllCaps;      // Turn on the 'all caps' field 
        }

        int result = (int)(StatementAttributes.AllOn & attributes);
        string bobResponse;

        switch (result)
        {
            case 1:
                bobResponse = "Fine. Be that way!";
                break;
            case 2:
                bobResponse = "Sure.";
                break;
            case 4:
                bobResponse = "Whoa, chill out!";
                break;
            case 6:
                bobResponse = "Calm down, I know what I'm doing!";
                break;
            default:
                bobResponse = "Whatever.";
                break;
        }
        return bobResponse;
    }

    // check if input string has any letters. if not, return false. if so, check for all uppercase.
    static bool IsAllCaps(string input)
    {
        bool hasLetter = input.Any(index => Char.IsLetter(index));

        foreach (char c in input)
        {
            if (Char.IsLetter(c) && !Char.IsUpper(c))
            {
                    return false;
            }
        }
        return true && hasLetter;
    }

    // trim string and check if there are any characters left. trim removes \n, \r, \t and outer spaces
    static bool IsEmpty(string input)
    {
        return input.Trim().Length == 0;
    }

    // trim end of string and check if the last charactre is a question mark
    static bool IsQuestion(string input)
    {
        string trimmed = input.TrimEnd();
        return trimmed.Length > 0 ? trimmed[trimmed.Length - 1].Equals('?') : false;
    }
}

File: JavaScript/bob.js
export const hey = (message = '') => {
  let isEmpty, isQuestion, isAllCaps, lettersOnly, result, trimmed, response = "", types = [];
  trimmed = message.trim();

  // determine what type of message
  isEmpty = trimmed.length === 0;
  isQuestion = !isEmpty ? trimmed[trimmed.length - 1] === '?' : false;      // last char is ?
  lettersOnly = !isEmpty ? trimmed.match(/[a-zA-Z]/g) : '';                 // filters only letters. if input is numbers only produces null
  lettersOnly = lettersOnly != null && lettersOnly.length > 0 ? lettersOnly.join('') : '';    // if not null or empty, join as string
  isAllCaps = lettersOnly.length > 0 && lettersOnly === lettersOnly.toUpperCase();  // determines if has letters and all uppercase

  // types [empty, question, caps, all, initial state]
  types = [1, 2, 4, 7, 0];

  // uses bitwise | operator to turn on bits for appropriate message type
  let messageAttributes = types[4];
  isEmpty ? messageAttributes |= types[0] : null;      // if empty, turn on empty field, else do nothing
  isQuestion ? messageAttributes |= types[1] : null;   // if question, turn on question field, else do nothing
  isAllCaps ? messageAttributes |= types[2] : null;    // if all caps, turn on all caps field, else do nothing

  // uses bitwise & operator to compare each position to see what is a 1
  result = types[3] & messageAttributes;
  switch (result) {
    case 1:
      response = "Fine. Be that way!";
      break;
    case 2:
      response = "Sure.";
      break;
    case 4:
      response = "Whoa, chill out!";
      break;
    case 6:
      response = "Calm down, I know what I'm doing!";
      break;
    default:
      response = "Whatever.";
      break;
  }
  return response;
};